### PR TITLE
Feature/activity log

### PR DESCRIPTION
**Pull request by AhmedMahdy14**

### CODE REVIEW

File: apihub_users/subscription/depends.py [hunk 1]
```diff
@@ -12,52 +13,35 @@
...
-    return username
+    return SubscriptionBase(
+        username=username, tier=tier_claim, application=subscription_claim
+    )
 
 
 def require_subscription_balance(
-    application: str,
-    username: str = Depends(require_subscription),
+    subscription_base: SubscriptionBase = Depends(require_subscription),
```
Comment (yifan) [inline apihub_users/subscription/depends.py:36]: lets call this variable subscription, it will look better

File: tests/test_activity.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    yield TestClient(app)
+
+
+class TestStatistics:
+    def test_create_activity(self, db_session):
+        ActivityFactory._meta.sqlalchemy_session = db_session
+        ActivityFactory._meta.sqlalchemy_session_persistence = "commit"
+
+        assert (
+            ActivityQuery(db_session).create_activity(
```
Comment (yifan) [inline tests/test_activity.py:69]: To test a database operation, we can call the function and verify the result in database. We need to make sure the result (side effect) is correct.

File: tests/test_activity.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+                    ip_address="",
+                    latency=0.0,
+                )
+            )
+            is None
+        )
+
+    def test_get_activity_by_key(self, client, db_session):
+        query = ActivityQuery(db_session)
+        assert query.get_activity_by_key("app1_key") is not None
```
Comment (yifan) [inline tests/test_activity.py:87]: same here, verify the result

File: tests/test_subscription.py [hunk 4]
```diff
@@ -9,8 +9,6 @@
 from apihub_users.security.schemas import UserBase, UserType
 from apihub_users.security.depends import require_user, require_admin, require_token
 from apihub_users.subscription.depends import (
-    require_subscription,
-    update_subscription_balance,
     require_subscription_balance,
 )
 from apihub_users.subscription.models import Subscription, SubscriptionType
```
Comment (yifan) [inline tests/test_subscription.py:14]: We can change SubscriptionType to SubscriptionTier to match the tier modificaiton

**Review by yifan — COMMENTED** [review]
minor changes needed

**Comment by yifan** [comment]
@AhmedMahdy14 unittest does not pass

File: apihub_users/subscription/depends.py (hunk 1 above)
Comment (AhmedMahdy14) [reply apihub_users/subscription/depends.py:36]: ok.

**Pull request closed without merging**